sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala

```diff
@@ -175,10 +175,30 @@ class UnivocityParser(
       }
 
     case _: TimestampType => (d: String) =>
-      nullSafeDatum(d, name, nullable, options)(timestampFormatter.parse)
+      nullSafeDatum(d, name, nullable, options) { datum =>
+        try {
+          timestampFormatter.parse(datum)
+        } catch {
+          case NonFatal(e) =>
+            // If fails to parse, then tries the way used in 2.0 and 1.x for backwards
+            // compatibility.
+            val str = UTF8String.fromString(datum)
+            DateTimeUtils.stringToTimestamp(str, options.zoneId).getOrElse(throw e)
```

Comment:
Looking at what was removed there https://github.com/apache/spark/pull/23150/files#diff-c82e4b74d2a51fed29069745ce4f9e96L164 , `DateTimeUtils.stringToTime` and `DateTimeUtils.stringToTimestamp` are 2 different functions, see https://github.com/MaxGekk/spark/blob/60c0974261c947c0838457c40f4fe0e64d17ca15/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala#L173-L191 

`stringToTime` has a few problems - it doesn't respect to Spark's session time zone, and parses in the combined calendar. If you use it as a fallback, you can get parsed values in different calendars.

Reply:
Yea I just want to support the legacy format, not to be exactly the same with 2.4. We switch the calendar so we can't be exactly the same anyway.

Reply:
If we can't be exactly the same, why do we add this back?
cc @marmbrus

Reply:
We can't be exactly the same in 1% cases, and this change can make us exactly the same in 99% cases.

Reply:
- What cases used to be supported that are no longer supported?
 - What remains in the 1%?
 - Why can we not get 100%?

To me, Spark suddenly failing to parse timestamps that it used to parse fall into the "extremely annoying" category of regressions. Working with timestamps is already hard and confusing. If my job just suddenly starts returning `null` when it used to work as expected its:
 - Probably going to take a while for people to figure it out (while the `null` partition of their table mysteriously fills up).
 - Probably non-trival for them to figure out how to fix it. (I don't think constructing / testing format strings is a trivial activity).

I really question why we changed these API's rather than creating new better ones. These changes broke my pipeline and it took me hours + access to a team of spark committers to figure out why!!!!!

Reply:
We switch the calendar from "hybrid calendar (Julian + Gregorian)" to "ISO chronology (Proleptic Gregorian)", which affects date/timestamp values before October 15, 1582. For example, some dates are invalid according to the new calendar but the old calendar accepts it. When we add days to a date, the calendar skips invalid dates.

The new calendar is widely used by RDBMS and others like parquet, snowflake, but we should have a better rollout plan for the calendar switch, like making the calendar configurable for Spark. Unfortunately, we catch this too late.

However, the more serious problem is the switch of the pattern string(as we switch the JDK datetime API internally), and 3.0 can't parse some date/timestamp strings suddenly. This is the first step to fix this issue by adding back the legacy fallback, we are still working on fixing the 1% when users set a custom pattern string:
1. try to support the previous pattern string with the new calendar. If can't, go 2)
2. detect the conflicts between old and new pattern string and throw an exception to tell users what to do (turn on the legacy config or accept the new pattern string knowing what it mean)

Reply:
I agree with your analysis of the relative costs here.

I think there are probably very few Spark programs in existence where the specifics of date handling for dates before October 15, 1582 are critical to correctness. This is especially true if you still get back some valid date (rather than `null`). If thats really the only difference with the calendar switch, then I think you could even make an argument that making the calendar configurable is too high a cost for the project.

Removing support for previously supported date format parsing strings in contrast seems pretty disastrous to me. Having to remove test cases really should have been a red flag to reviewers of those changes that we were going to break users' programs.

Reply:
My five cents, besides of the calendar problem the fallback function `DateTimeUtils.stringToTime()` didn't take into account the session time zone (`spark.sql.session.timeZone`). This can lead to the situation when there are values in a column parsed in different time zones. I do think this is serious problem of the fallback mechanism. If we are going to restore it to only don't get nulls, this will make users life harder in troubleshooting wrong results.

Reply:
@MaxGekk now we call `DateTimeUtils.stringToTimestamp(str, options.zoneId)`, so timezone is not an issue?

Reply:
`stringToTimestamp` doesn't have the issue but you implemented different fallback mechanism. From my point of view, it makes sense to either:
- restore the old function `stringToTime`
- or keep current code as is.

Fallback to `stringToTimestamp` is new feature, I think.

Reply:
I know they are different implementations, but we care more about the behavior to end users.

Are there any timestamp strings that can be parsed by `stringToTime` but not `stringToTimestamp`?

Reply:
`stringToTimestamp` cannot parse time zone id with the `GMT` prefix, see https://github.com/MaxGekk/spark/blob/60c0974261c947c0838457c40f4fe0e64d17ca15/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala#L176

Reply:
looks like a corner case, but let's add it back as well.

Reply:
@cloud-fan Here is a draft PR which supports any zone ids by `stringToTimestamp`: https://github.com/apache/spark/pull/27753

Reply:
Thank you for keeping moving this situation!

Reply:
@MaxGekk that's a good complement to support more legacy formats!